File: Layer.py
try:
    import mathlib as Mathlib   #< c++ version (much faster training). Must be compiled first
except ImportError:
    import Mathlib
import Activation
import math

class Layer:
    """A collection of neurons forming a single layer in the network"""
    def __init__(self, inputCount: int, neuronCount: int,
                 activationFunc: Activation=Activation.Pass,   #< Neuron-level activation function
                 ):
        
        self.inputsLen = inputCount
        self.activationFunc = activationFunc
        self.weights, self.biases = self._createLayerData(inputCount, neuronCount)

        self.out = []
        self.preActivationOut = []
        self.inputs = []
        self.d_inputs = []
        self.d_weights = []
        self.d_biases = []

    def _createLayerData(self, inputCount: int=10, neuronCount: int=5) -> tuple[list[float], list[float]]:
        if self.activationFunc in (Activation.ReLU, Activation.LeakyReLU):
            scale = 2.0 / math.sqrt(inputCount)  #< He/Kaiming initialization, scale by 2/sqrt(inputCount) to prevent overflow (better for ReLu activation functions)
        else:
            scale = 1.0 / math.sqrt(inputCount)  #< Xavier/Glorot initialization, scale by 1/sqrt(inputCount) to prevent overflow (better for sigmoid activation functions)
        weights = [[Mathlib.randomNumber(-scale, scale, 2) for _ in range(inputCount)] for _ in range(neuronCount)]
        biases = [0.0 for _ in range(neuronCount)]  # Initialize biases to 0
        return(weights, biases)

    def forward(self, inputs: list[float]) -> list[float]:
        """
        Compute the neurons' output.
        Multiply each input by its corresponding weight, sum the
        results, add the bias and then apply the activation function.
        
        multiply the inputs [i1, i2.. in] with weights [w1, w1... wn] to get i1*w1+i2*w2... +in*wn, then add the bias
        """
        self.inputs = inputs
        if len(self.weights[0]) != len(self.inputs):
            raise Exception("inputs and weight must have the same length!")
        self.preActivationOut = [Mathlib.dot2Vectors(self.inputs, weights)+bias for weights, bias in zip(self.weights, self.biases)]     #< sum(w*x)+b
        self.out = [self.activationFunc.forward(out) for out in self.preActivationOut]
        return(self.out)

    def backward(self, d_values:list[float], inputs=None, preActivationOut=None) -> list[float]:
        """ Compute gradient 
        The following is the explanation for the backward for a SINGLE neuron, a layer would just be a list of these.
            since forward is computed as Activation(sum(x1*w1, x2*w2..., bias)),
            the backward for the weights would be computed as [Activation`(sum(...))*sum`(...)*xi] for every element,
            and the backward for the inputs would be computed as [Activation`(sum(...))*sum`(...)*wi] for every element,
            note that sum`(...) is equal to 1 no matter the reference or input.
            this makes backward for the weights be simplified to [Activation`(sum(...))*xi] for every element,
            and the backward for the inputs be [Activation`(sum(...))*wi] for every element.
        """
        if inputs == None:
            inputs = self.inputs
        if preActivationOut == None:
            preActivationOut = self.preActivationOut
        self.d_inputs = Mathlib.zeroes(len(inputs))
        self.d_weights = []
        self.d_biases = []
        for out, weights, d_val in zip(preActivationOut, self.weights, d_values):
            activation_dx = self.activationFunc.backward(out)*d_val      #< chain rule
            self.d_inputs = Mathlib.addTwoVectors([activation_dx*w for w in weights], self.d_inputs)   #< Total derivative is the sum of the derivatives with respect to the input of each neuron since they are all the same inputs
            self.d_weights.append([activation_dx*i for i in inputs])
            self.d_biases.append(activation_dx)

        return(self.d_inputs)

    def getWeights(self) -> list[list[float]]:
        """ return a list of lists of all weights """
        return(self.weights)
    def setWeights(self, weights: list[list[float]]) -> None:
        self.weights = weights
    def addToWeight(self, indexRow: int, indexCol: int, val: float) -> None:
        self.weights[indexRow][indexCol] = val

    def getBiases(self) -> list[float]:
        """ return a list of all biases """
        return(self.biases)
    def setBiases(self, biases: list[float]) -> None:
        self.biases = biases
    def addToBias(self, index: int, val: float) -> None:
        self.biases[index] = val

    def getOutputs(self) -> list[float]:
        return(self.out)
    def getPreActivationOutputs(self) -> list[float]:
        return(self.preActivationOut)

    def getInputGradient(self) -> list[float]:
        return(self.d_inputs)
    def getWeightsGradient(self) -> list[list[float]]:
        return(self.d_inputs)
    def getBiasesGradient(self) -> list[float]:
        return(self.d_inputs)

File: CPP_Network/Layer.cpp
#include <cmath>
#include <vector>
#include <stdexcept>

#include "Layer.hpp"
#include "Types.hpp"
#include "Mathlib.hpp"      // My custom mathlib
#include "Activation.hpp"   // My custom Actiavtion Classes

Layer::Layer(int inputCount, int neuronCount, ActivationType type) {
    this->inputCount = inputCount;
    this->neuronCount = neuronCount;

    this->d_inputs = zeroes(this->inputCount);
    this->_setActivation(type);
    this->_createWeights();
    this->_createBiases();
    this->_createOutputs();
}

void Layer::_setActivation(ActivationType type) {
        switch (type) {
            case ActivationType::RELU:
                forwardCallback = &ReLU::forward;
                backwardCallback = &ReLU::backward;
                break;
            case ActivationType::LEAKY_RELU:
                forwardCallback = &LeakyReLU::forward;
                backwardCallback = &LeakyReLU::backward;
                break;
            case ActivationType::PASS:
            default:
                forwardCallback = &Pass::forward;
                backwardCallback = &Pass::backward;
                break;
        }
    }

void Layer::_createWeights() {
    double scale;
    scale = 2.0 / std::sqrt(this->inputCount);  // He/Kaiming initialization, scale by 2/sqrt(inputCount) to prevent overflow (better for ReLu activation functions)

    this->weights = std::vector<std::vector<double>>(this->neuronCount, std::vector<double>(this->inputCount, 0.0));
    this->d_weights = this->weights;
    for (int i = 0; i < this->neuronCount; i++) {
        for (int j=0; j < this->inputCount; j++) {
            this->weights[i][j] = randomNumber(-scale, scale, 2);
        }
    }
}

void Layer::Layer::_createBiases() {
    this->biases = std::vector<double>(this->neuronCount, 0.0);
    this->d_biases = this->biases;
}

void Layer::Layer::_createOutputs() {
    this->out = std::vector<double>(this->neuronCount, 0.0);
    this->preActivationOut = this->out;
}

std::vector<double> Layer::forward(const std::vector<double>& inputs) { // inputs are not mutated
    /*
    Compute the neurons' output.
    Multiply each input by its corresponding weight, sum the
    results, add the bias and then apply the activation function.
    
    multiply the inputs [i1, i2.. in] with weights [w1, w1... wn] to get i1*w1+i2*w2... +in*wn, then add the bias
    */
    this->inputs = inputs;
    if (this->weights[0].size() != this->inputs.size()) {
        throw std::invalid_argument("inputs and weight must have the same length!");
    }
    for (int i=0; i < this->weights.size(); i++ ) {
        this->preActivationOut[i] = dot2Vectors(this->inputs, this->weights[i])+this->biases[i]; // sum(w*x)+b
        this->out[i] = this->forwardCallback(this->preActivationOut[i]);
    }
    return this->out;
}

std::vector<double> Layer::backward(const std::vector<double>& d_values, const std::vector<double>& inputs, const std::vector<double>& preActivationOut) {
    std::fill(this->d_inputs.begin(), this->d_inputs.end(), 0.0);
    for (int n=0; n<this->neuronCount; n++) {
        double activation_dx = this->backwardCallback(preActivationOut[n]) * d_values[n];      // chain rule
        for (int i=0; i<this->inputCount; i++ ) {
            this->d_inputs[i] = activation_dx*this->weights[n][i]+this->d_inputs[i];   // chain rule
            this->d_weights[n][i] = activation_dx*inputs[i]; 
        }
        this->d_biases[n] = activation_dx;
    }

    return d_inputs;
}

std::vector<double> Layer::backward(const std::vector<double>& d_values) {
    /*
    Compute gradient 
    The following is the explanation for the backward for a SINGLE neuron, a layer would just be a list of these.
        since forward is computed as Activation(sum(x1*w1, x2*w2..., bias)),
        the backward for the weights would be computed as [Activation`(sum(...))*sum`(...)*xi] for every element,
        and the backward for the inputs would be computed as [Activation`(sum(...))*sum`(...)*wi] for every element,
        note that sum`(...) is equal to 1 no matter the reference or input.
        this makes backward for the weights be simplified to [Activation`(sum(...))*xi] for every element,
        and the backward for the inputs be [Activation`(sum(...))*wi] for every element.
    */
    return this->backward(d_values, this->inputs, this->preActivationOut);
}

std::vector<std::vector<double>> Layer::getWeights() {
    return this->weights;
}
void Layer::setWeights(std::vector<std::vector<double>> weights) {
    this->weights = weights;
}
void Layer::addToWeight(int indexRow, int indexCol, double val) {
    this->weights[indexRow][indexCol] += val;
}

std::vector<double> Layer::getBiases() {
    return this->biases;
}
void Layer::setBiases(std::vector<double> biases) {
    this->biases = biases;
}
void Layer::addToBias(int index, double val) {
    this->biases[index] += val;
}

std::vector<double> Layer::getOutputs() {
    return this->out;
}

std::vector<double> Layer::getPreActivationOutputs() {
    /* The output of the layer before passing the activation function */
    return this->preActivationOut;
}

std::vector<double> Layer::getInputGradient() {
    return this->d_inputs;
}

std::vector<std::vector<double>> Layer::getWeightsGradient() {
    return this->d_weights;
}

std::vector<double> Layer::getBiasesGradient() {
    return this->d_biases;
}